PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/LinkAzureServiceBeforeRunProvider.java

```diff
@@ -0,0 +1,192 @@
+        if (MapUtils.isNotEmpty(linkedEnvMap)) {
+            // set envs for remote deploy
+            if (runConfiguration instanceof WebAppConfiguration) {
+                IntelliJWebAppSettingModel model = ((WebAppConfiguration) runConfiguration).getModel();
+                AzureWebAppMvpModel.getInstance().updateWebAppSettings(model.getSubscriptionId(), model.getWebAppId(), linkedEnvMap, new HashSet<>());
+            }
+            // set envs for local run
+            if (runConfiguration instanceof AbstractRunConfiguration
+                    || StringUtils.equals(runConfiguration.getClass().getName(), SPRING_BOOT_CONFIGURATION_REF)
+                    || runConfiguration instanceof ApplicationConfiguration) {
+                ((ModuleBasedConfiguration<?, ?>) runConfiguration).putUserData(LINK_AZURE_SERVICE, true);
+                ((ModuleBasedConfiguration<?, ?>) runConfiguration).putUserData(LINK_AZURE_SERVICE_ENVS, linkedEnvMap);
+            }
+        }
+        return true;
+    }
+
+    private String getModuleName(@NotNull RunConfiguration runConfiguration) {
+        if (runConfiguration instanceof WebAppConfiguration) {
+            String name = ((WebAppConfiguration) runConfiguration).getName();
+            name = name.substring(name.indexOf(":") + 1);
+            name = name.substring(name.indexOf(":") + 1);
+            return name;
+        }
+        if (runConfiguration instanceof AbstractRunConfiguration
+                || StringUtils.equals(runConfiguration.getClass().getName(), SPRING_BOOT_CONFIGURATION_REF)) {
+            RunConfigurationModule module = ((ModuleBasedConfiguration<?, ?>) runConfiguration).getConfigurationModule();
+            return module.getModule().getName();
+        }
+        if (runConfiguration instanceof ApplicationConfiguration) {
+            return ((ApplicationConfiguration) runConfiguration).getConfigurationModule().getModule().getName();
+        }
+        return null;
+    }
+
+    private void retrieveEnvMap(Project project, Map<String, String> linkedEnvMap, String moduleName) {
+        List<LinkPO> moduleRelatedLinkerList = AzureLinkStorage.getProjectStorage(project).getLinkersByModuleId(moduleName)
+                .stream()
+                .filter(e -> LinkType.SERVICE_WITH_MODULE.equals(e.getType()))
+                .collect(Collectors.toList());
+        if (CollectionUtils.isEmpty(moduleRelatedLinkerList)) {
+            return;
+        }
+        // services in application level
+        Set<? extends BaseServicePO> serviceSet = AzureMySQLStorage.getStorage().getServices();
+        for (BaseServicePO service : serviceSet) {
+            for (LinkPO linker : moduleRelatedLinkerList) {
+                if (!StringUtils.equals(linker.getServiceId(), service.getId())) {
+                    continue;
+                }
+                String envPrefix = linker.getEnvPrefix();
+                if (ServiceType.AZURE_DATABASE_FOR_MYSQL.equals(service.getType())) {
```

Comment:
`LinkAzureServiceBeforeRunProvider` should link all azure service to project instead of Azure MySQL only, and I think we may make service to generate correspond env map and before run task will only collect result from all linked services and return.

Reply:
extractly correct.